rfcs/20201103-binary-bcast-source0/README.md

```diff
@@ -0,0 +1,43 @@
+# Proposal for enable broadcast source 0 in binary
+
+## Problem and motivation
+Currently in OneDNN library binary primitive supports implicit broadcast 
+semantics for source 1.It means that if some dimension has value of one, 
+this value will be used to compute an operation with each point of 
+source 0 for this dimension. There is also a need for enable this functionality
+for source 1 to handle binary operation on sources like `Nx1` and `1xM`. This 
+will help better support OneDNN in TensorFlow library. At this point, was 
+reported, that performance(jit kernel) is not the main concern.
+
+![](bcast_example.png "Binary ADD operation 1x4:4x1")
+
+## Overview
+
+### Current implementations
+*  Reference implementation allows broadcast with following syntax:
+   `NxCxDxHxW:N/1xC/1xD/1xH/1xW/1 -> NxCxDxHxW`
+*  JIT implementation allows broadcast with following syntax:
+   `NxCxDxHxW:{NxCx1x1x1,1xCx1x1x1,1x1x1x1x1} -> NxCxDxHxW`
+*  JIT i8 implementation allows broadcast with following syntax:
+   `NxCxDxHxW:{N,1}xCx1x1x1 -> NxCxDxHxW`
+### New implementations
+*  Reference implementation allows broadcast with following syntax:
+   `N/1xC/1xD/1xH/1xW/1:N/1xC/1xD/1xH/1xW/1 -> NxCxDxHxW`
+### Possible future implementations (if the need is reported)
+*  JIT implementation allows broadcast with following syntax:
+   `{NxCxDxHxW,NxCx1x1x1,1xCx1x1x1,1x1x1x1x1}:
+    {NxCxDxHxW,NxCx1x1x1,1xCx1x1x1,1x1x1x1x1} -> NxCxDxHxW`
+*  JIT i8 implementation allows broadcast with following syntax:
+   `{N,1}xCx{D,1}x{H,1}x{W,1}:{N,1}xCx{D,1}x{H,1}x{W,1} -> NxCxDxHxW`
+
+Similarly for other formats.
+
+## Proposal
+The proposal is to modify reference algorithm to enable any broadcast on 
+source 0 and source 1. The dimensions of the destination should be defined as: 
+`dst_dims[d] = max(src0_dims[d], src1_dims[d])` and 
+`src0_dims[d] == src1_dims[d] || one_of(1, src0_dims[d], src1_dims[d])`. Thanks 
+to this, we will get backward compatibility. If user doesnt use broadcast on 
+source0, source0 dimensions are the same as dst. The post-ops operations also 
+remains unchanged. If it will be reported, can also add support for broadcast 0
+in jit kernels.
```

Comment:
@bartekxk Thanks for the RFC!
In general proposal looks good, the only question is what to do with binary post operation? I think we should investigate this part deeper, because it might affect both developers and users in the future, so it should be at least documented. There are a few options:

1. Add a limitation to binary post operation so dst_dims should be equal to src0_dims.
    - Pros:
        - simple.
    - Cons:
         - complicates integration into an application. Before this change an application could match any combination of         some primitive and a binary primitive and fuse them together without looking into shapes. After this change an         application should track binary sizes and create a fusion only when src0_dims are equal to dst_dims.
2. Extend binary post operation to support src0 broadcast.
    - Pros:
        - ease of use (see cons of the first option).
    - Cons:
        - complicates post operations a lot:
            - post operation changes destination sizes of a primitive it is a part of.
            - I believe implementation will be very complicated due to broadcast of src0 and will require additional buffers which will limit performance benefit from binary post operation.

There are might be other options I am not aware of, so please don't take these options as **only** options.

Reply:
In general there are probably only two options - support or not. One more thing we can consider is for example jit or ref implementation broadcast source 0 in post-ops. I mentioned about it in second option and I think it can introduce a lot of complications in documentation. I added your options with some changes. Thanks.